helix-core/src/main/java/org/apache/helix/api/rebalancer/waged/constraints/SoftConstraint.java

```diff
@@ -0,0 +1,49 @@
+package org.apache.helix.api.rebalancer.waged.constraints;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.helix.controller.rebalancer.waged.model.AssignableNode;
+import org.apache.helix.controller.rebalancer.waged.model.AssignableReplica;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterContext;
+
+/**
+ * Evaluate a partition allocation proposal and return a score within the normalized range.
+ * A higher score means the proposal is more preferred.
+ */
+public interface SoftConstraint {
```

Comment:
Suggest using abstract class instead of interface. Because the constraint importance is a must-have required field, abstract class could have the default constructor.

```
public abstract class SoftConstraint {
   private float _weight;
   SoftConstraint(float weight) {
     _weight = weight;
   }
   ...
}
```

Reply:
We do have an abstract class. But not in this initial checking. That will be in our implementation.
The interface is better for flexibility.